sdk/monitor/azure-monitor-ingestion/src/main/java/com/azure/monitor/ingestion/implementation/Utils.java

```diff
@@ -0,0 +1,183 @@
+        if (ex.getValue() instanceof LinkedHashMap<?, ?>) {
+            @SuppressWarnings("unchecked")
+            LinkedHashMap<String, Object> errorMap = (LinkedHashMap<String, Object>) ex.getValue();
+            if (errorMap.containsKey("error")) {
+                Object error = errorMap.get("error");
+                if (error instanceof LinkedHashMap<?, ?>) {
+                    @SuppressWarnings("unchecked")
+                    LinkedHashMap<String, String> errorDetails = (LinkedHashMap<String, String>) error;
+                    if (errorDetails.containsKey("code") && errorDetails.containsKey("message")) {
+                        responseError = new ResponseError(errorDetails.get("code"), errorDetails.get("message"));
+                    }
+                }
+            }
+        }
+        return responseError;
+    }
+
+    /**
+     * Gzips the input byte array.
+     * @param bytes The input byte array.
+     * @return gzipped byte array.
+     */
+    public static byte[] gzipRequest(byte[] bytes) {
+        // This should be moved to azure-core and should be enabled when the client library requests for gzipping the
+        // request body content.
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        try (GZIPOutputStream zip = new GZIPOutputStream(byteArrayOutputStream)) {
+            zip.write(bytes);
+        } catch (IOException exception) {
+            throw LOGGER.logExceptionAsError(new UncheckedIOException(exception));
+        }
+        return byteArrayOutputStream.toByteArray();
+    }
+
+    public static void writeLogsAndCloseJsonGenerator(JsonGenerator generator, List<String> serializedLogs) throws IOException {
+        generator.writeRaw(serializedLogs.stream()
+            .collect(Collectors.joining(",")));
+        generator.writeEndArray();
+        generator.close();
+    }
+
+    public static ObjectSerializer getSerializer(LogsUploadOptions options) {
+        if (options != null && options.getObjectSerializer() != null) {
+            return options.getObjectSerializer();
+        }
+
+        return  DEFAULT_SERIALIZER;
+    }
+
+    public static int getConcurrency(LogsUploadOptions options) {
+        if (options != null && options.getMaxConcurrency() != null) {
+            return options.getMaxConcurrency();
+        }
+
+        return  1;
+    }
+
+    /**
+     * Creates cached (that supports scaling) thread pool with shutdown hook to do best-effort graceful termination within timeout.
+     * @param timeoutSec Timeout in seconds to wait for tasks to complete or terminate.
+     * @return {@link ExecutorService} instance.
+     */
+    public static ExecutorService getThreadPoolWithShutDownHook(int timeoutSec) {
+        ExecutorService threadPool = Executors.newCachedThreadPool();
+        registerShutdownHook(threadPool, timeoutSec);
```

Comment:
What happens to requests that are made to the client after the threadpool is shutdown?

Reply:
they're rejected: 

`assertThrows(RejectedExecutionException.class, () -> threadPool.submit(() -> stream.collect(Collectors.toList())));`

Reply:
For usecases where logs are rotated every hour, the client will attempt to upload logs after an hour causing rejected exception. This will make the sync client unusable if the client is idle for a while (timeout period) and a new instance has to be created.

Since we are using cached thread pool, I don't think calling shutdown is required. It doesn't keep any thread in the pool idle for more than 60 seconds (by default) and will let the JVM terminate when the application terminates. This will allow the client to accept new requests after any amount of idle time and create new threads when needed.

Also, we should make the sync client implement `AutoCloseable` and shut down the ExceutorService when `close()` is called. So, if customers want to close the app immediately (and not wait 60 secs for any active thread in the pool to be terminated), they can close the client.

Reply:
per offline conversation made thread pool an instance variable and shutting it down when client is closed (plus adding the hook in case user forgets to close it).